// @cloudinary/vue/index.ssr.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var html = require('@cloudinary/html');
var vue = require('vue');
var serverRenderer = require('vue/server-renderer');

/**
 * Used in production
 */
const SDKAnalyticsConstants = {
    sdkSemver: "1.11.2",
    techVersion: "3.2.37",
    sdkCode: "L",
};

// Disabled linting due to [@vue/compiler-sfc] `defineProps` is a compiler macro and no longer needs to be imported.
// eslint-disable-next-line no-undef
/**
 * @mixin VueSDK
 * @description The Cloudinary Vue SDK contains components like \<AdvancedImage\> to easily render your media assets from Cloudinary.
 * The SDK also comes with support for optional JS plugins that make the components smart, with features like lazy loading, placeholder, accessibility & responsiveness.
 *
 * @example
 * <caption>
 *  Please note that the order of the plugins is important. See {@link https://cloudinary.com/documentation/sdks/js/frontend-frameworks/index.html#plugin-order|Plugin Order} for more details.
 * </caption>
 * // AdvancedImage
 * <template>
 *   <div>
 *     <AdvancedImage :cldImg="cldImg" :plugins="plugins" />
 *   </div>
 * </template>
 *
 * <script lang="ts">
 * import { defineComponent } from 'vue';
 * import { CloudinaryImage } from '@cloudinary/url-gen/assets/CloudinaryImage';
 * import { AdvancedImage, responsive } from '@cloudinary/vue';
 *
 * export default defineComponent({
 *   name: 'App',
 *   components: {
 *     AdvancedImage,
 * },
 *   setup(props) {
 *     const cldImg = new CloudinaryImage(
 *       'sample',
 *       { cloudName: 'demo' },
 *       { analytics: false }
 *   );
 *
 *     const plugins = [responsive({ steps: 100 })];
 *
 *     return {
 *       cldImg,
 *       plugins,
 *   };
 *  },
 * });
 * </ script>
 *
 * @example
 * <caption>
 *  Using `AdvancedVideo` custom defined resources.
 * </caption>
 *
 * // AdvancedVideo
 * <template>
 *   <AdvancedVideo :cldVid="cldVid" :sources="sources" controls width="600" />
 * </template>
 *
 * <script lang="ts">
 * import { defineComponent } from "vue";
 * import { auto } from "@cloudinary/url-gen/qualifiers/videoCodec";
 * import { videoCodec } from "@cloudinary/url-gen/actions/transcode";
 * import { AdvancedVideo } from "../dist";
 * import { CloudinaryVideo } from "@cloudinary/url-gen/assets/CloudinaryVideo";
 *
 * export default defineComponent({
 *   name: "App",
 *   components: {
 *     AdvancedVideo,
 *  },
 *   setup(props) {
 *     const cldVid = new CloudinaryVideo(
 *       "dog",
 *       { cloudName: "demo" },
 *       { analytics: false }
 *     );
 *
 *     const sources = [
 *       {
 *         type: "mp4",
 *         transcode: videoCodec(auto()),
 *       },
 *       {
 *         type: "webm",
 *         transcode: videoCodec(auto()),
 *       },
 *     ];
 *
 *     return {
 *       cldVid,
 *       sources,
 *     };
 *   },
 * });
 * </ script>
 */
/**
 * @memberOf VueSDK
 * @module AdvancedImage
 * @description The Cloudinary image component.
 * @vue-prop {CloudinaryImage} cldImg Generated by @cloudinary/url-gen
 * @vue-prop {Plugins} plugins Advanced image component plugins accessibility(), responsive(), lazyload(), placeholder()
 */
var script$1 = /*#__PURE__*/ vue.defineComponent({
    __name: 'AdvancedImage',
    __ssrInlineRender: true,
    props: {
        cldImg: { type: null, required: true },
        plugins: { type: null, required: false }
    },
    setup(__props) {
        const props = __props;
        const imageRef = vue.ref(null);
        let htmlLayerInstance;
        const getSsrSrc = () => html.serverSideSrc(props.plugins, props.cldImg, SDKAnalyticsConstants);
        /**
         * On mount, creates a new HTMLLayer instance and initializes with ref to img element,
         * user generated cloudinaryImage and the plugins to be used.
         */
        vue.onMounted(() => {
            htmlLayerInstance = new html.HtmlImageLayer(imageRef.value, props.cldImg, props.plugins, SDKAnalyticsConstants);
        });
        /**
         * On update, we cancel running plugins and update image instance with the state of user
         * cloudinaryImage and the state of plugins.
         */
        vue.onUpdated(() => {
            html.cancelCurrentlyRunningPlugins(htmlLayerInstance.htmlPluginState);
            // call html layer to update the dom again with plugins and reset toBeCanceled
            htmlLayerInstance.update(props.cldImg, props.plugins, SDKAnalyticsConstants);
        });
        /**
         * On unmount, we cancel the currently running plugins.
         */
        vue.onUnmounted(() => {
            // Safely cancel running events on unmount.
            html.cancelCurrentlyRunningPlugins(htmlLayerInstance.htmlPluginState);
        });
        return (_ctx, _push, _parent, _attrs) => {
            if (vue.unref(html.isBrowser)()) {
                _push(`<img${serverRenderer.ssrRenderAttrs(vue.mergeProps({
                    ref_key: "imageRef",
                    ref: imageRef
                }, _attrs))}>`);
            }
            else {
                _push(`<img${serverRenderer.ssrRenderAttrs(vue.mergeProps({
                    src: getSsrSrc()
                }, _attrs))}>`);
            }
        };
    }
});

script$1.__file = "src/components/AdvancedImage.vue";

// Disabled linting due to [@vue/compiler-sfc] `defineProps` is a compiler macro and no longer needs to be imported.
// eslint-disable-next-line no-undef
/**
 * @memberOf VueSDK
 * @module AdvancedVideo
 * @description The Cloudinary video component.
 *
 * @vue-prop {CloudinaryVideo} cldVid Generated by @cloudinary/url-gen
 * @vue-prop {VideoPoster} cldPoster Generated by @cloudinary/url-gen
 * @vue-prop {Plugins} plugins Advanced video component plugins accessibility(), responsive(), lazyload(), placeholder()
 * @vue-prop {VideoSources} sources Optional sources to generate
 */
var script = /*#__PURE__*/ vue.defineComponent({
    __name: 'AdvancedVideo',
    __ssrInlineRender: true,
    props: {
        cldVid: { type: null, required: true },
        plugins: { type: null, required: false },
        sources: { type: null, required: false },
        cldPoster: { type: null, required: false }
    },
    setup(__props) {
        const props = __props;
        const videoRef = vue.ref(null);
        let htmlLayerInstance;
        /**
         * On mount, creates a new HTMLLayer instance and initializes with ref to img element,
         * user generated cloudinaryVideo and the plugins to be used.
         */
        vue.onMounted(() => {
            htmlLayerInstance = new html.HtmlVideoLayer(videoRef.value, props.cldVid, props.sources, props.plugins, undefined, props.cldPoster);
        });
        /**
         * On update, we cancel running plugins and update image instance with the state of user
         * cloudinaryVideo and the state of plugins.
         */
        vue.onUpdated(() => {
            html.cancelCurrentlyRunningPlugins(htmlLayerInstance.htmlPluginState);
            // call html layer to update the dom again with plugins and reset toBeCanceled
            htmlLayerInstance.update(props.cldVid, props.sources, props.plugins, undefined, props.cldPoster);
        });
        /**
         * On unmount, we cancel the currently running plugins.
         */
        vue.onUnmounted(() => {
            // Safely cancel running events on unmount.
            html.cancelCurrentlyRunningPlugins(htmlLayerInstance.htmlPluginState);
        });
        return (_ctx, _push, _parent, _attrs) => {
            _push(`<video${serverRenderer.ssrRenderAttrs(vue.mergeProps({
                ref_key: "videoRef",
                ref: videoRef
            }, _attrs))}></video>`);
        };
    }
});

script.__file = "src/components/AdvancedVideo.vue";

Object.defineProperty(exports, 'accessibility', {
  enumerable: true,
  get: function () { return html.accessibility; }
});
Object.defineProperty(exports, 'lazyload', {
  enumerable: true,
  get: function () { return html.lazyload; }
});
Object.defineProperty(exports, 'placeholder', {
  enumerable: true,
  get: function () { return html.placeholder; }
});
Object.defineProperty(exports, 'responsive', {
  enumerable: true,
  get: function () { return html.responsive; }
});
exports.AdvancedImage = script$1;
exports.AdvancedVideo = script;
